File: Raspi/Freenove_RFID_Starter_Kit_for_Raspberry_Pi/Code/Python_GPIOZero_Code/10.1.1_Nightlamp/ADCDevice.py
#!/usr/bin/env python3
########################################################################
# Filename    : ADCDevice.py
# Description : Freenove ADC Module library.
# Author      : www.freenove.com
# modification: 2020/04/21
########################################################################

import smbus

class ADCDevice(object):
    def __init__(self):
        self.cmd = 0
        self.address = 0
        self.bus=smbus.SMBus(1)
        # print("ADCDevice init")
        
    def detectI2C(self,addr):
        try:
            self.bus.write_byte(addr,0)
            print("Found device in address 0x%x"%(addr))
            return True
        except:
            print("Not found device in address 0x%x"%(addr))
            return False
            
    def close(self):
        self.bus.close()
        
class PCF8591(ADCDevice):
    def __init__(self):
        super(PCF8591, self).__init__()
        self.cmd = 0x40     # The default command for PCF8591 is 0x40.
        self.address = 0x48 # 0x48 is the default i2c address for PCF8591 Module.
        
    def analogRead(self, chn): # PCF8591 has 4 ADC input pins, chn:0,1,2,3
        value = self.bus.read_byte_data(self.address, self.cmd+chn)
        value = self.bus.read_byte_data(self.address, self.cmd+chn)
        return value
    
    def analogWrite(self,value): # write DAC value
        self.bus.write_byte_data(address,cmd,value)	

class ADS7830(ADCDevice):
    def __init__(self):
        super(ADS7830, self).__init__()
        self.cmd = 0x84
        self.address = 0x4b # 0x4b is the default i2c address for ADS7830 Module.   
        
    def analogRead(self, chn): # ADS7830 has 8 ADC input pins, chn:0,1,2,3,4,5,6,7
        value = self.bus.read_byte_data(self.address, self.cmd|(((chn<<2 | chn>>1)&0x07)<<4))
        return value

File: Raspi/Freenove_RFID_Starter_Kit_for_Raspberry_Pi/Libs/C-Libs/ADCDevice/ADCDevice.cpp
/**********************************************************************
* Filename    : ADCDevice.cpp
* Description : Freenove ADC Module library.
* Author      : www.freenove.com
* modification: 2020/03/06
**********************************************************************/
#include "ADCDevice.hpp"

ADCDevice::ADCDevice(){
    address = 0;
    wiringPiSetup(); //Initialize wiringPi.
}

int ADCDevice::detectI2C(int addr){
    _fd = wiringPiI2CSetup (addr);   
    if (_fd < 0){		
	printf("Error address : 0x%x \n",addr);
	return 0 ;
    } 
    else{	
	if(wiringPiI2CWrite(_fd,0) < 0){
	    printf("Not found device in address 0x%x \n",addr);
	    return 0;
	}
	else{
	    printf("Found device in address 0x%x \n",addr);
	    return 1 ;
	}
    }
}

int ADCDevice::analogRead(int chn){
    printf("Implemented in subclass! \n");
    return 0;
}


PCF8591::PCF8591(int addr)
{
    address = addr;
    cmd = 0x40;		//The default command for PCF8591 is 0x40.
    wiringPiSetup();
    detectI2C(address);
    printf("PCF8591 setup successful! \n");
}
int PCF8591::analogRead(int chn){
    wiringPiI2CWrite(_fd, cmd+chn);
    wiringPiI2CRead(_fd);
    wiringPiI2CWrite(_fd, cmd+chn);
    return wiringPiI2CRead(_fd);
}
int PCF8591::analogWrite(int value){
    return wiringPiI2CWriteReg8(_fd, cmd, value);
}

ADS7830::ADS7830(int addr)
{
    address = addr;
    cmd = 0x84;
    wiringPiSetup();
    detectI2C(address);
    printf("ADS7830 setup successful! \n");
}

int ADS7830::analogRead(int chn){
    wiringPiI2CWrite(_fd, cmd|(((chn<<2 | chn>>1)&0x07)<<4));
    return wiringPiI2CRead(_fd);
}
